content/courses/ada-idioms/chapters/controlling_obj_initialization_creation.rst

```diff
@@ -0,0 +1,1177 @@
+
+    private
+
+       type Contents is array (Positive range <>) of Element;
+
+       type Stack (Capacity : Positive) is limited record
+          Content : Contents (1 .. Capacity);
+          Top     : Natural :=0;
+       end record;
+
+       function Full (This : Stack) return Boolean is
+         (This.Top = This.Capacity);
+
+       function Empty (This : Stack) return Boolean is
+         (This.Top = 0);
+
+    end Bounded_Stacks;
+
+    package body Bounded_Stacks is
+
+       procedure Push (This : in out Stack; Value : Element) is
+       begin
+          This.Top := This.Top + 1;
+          This.Content (This.Top) := Value;
+       end Push;
+
+    end Bounded_Stacks;
+
+    with Ada.Text_IO;   use Ada.Text_IO;
+    with Bounded_Stacks; use Bounded_Stacks;
+
+    procedure Demo is
+       S : Stack (Capacity => 10);
+    begin
+       Push (S, 42);
+       Put_Line ("Done");
+    end Demo;
+
+The function :ada:`Empty` returns :ada:`True` when :ada:`Top` is zero, and zero
+is assigned to :ada:`Top` during default initialization. Consequently,
+:ada:`Assertion_Error` is not raised when :ada:`Demo` executes because the
+object :ada:`S` was indeed default initialized to the empty state.
+
+We said that when DIC is applied to a type, the run-time check is emitted for
+all object declarations of that type that rely on default initialization. But
+suppose the type does not define any default initialization. We can detect
+these uninitialized objects at run-time if we set the DIC Boolean expression
+to indicate that there is no default initialization defined for this type. The
+checks will then fail for those objects. That's the second solution to the
+initialization requirement.
+
+Specifically, we can express the lack of default initialization by a DIC
+condition that is hard-coded to the literal :ada:`False`. The evaluation during
+the check will then necessarily fail, raising :ada:`Assertion_Error`. Hence,
+for this type, explicit initialization is guaranteed in a program that does not
+raise :ada:`Assertion_Error` for this cause.
+
+The following is an example of the DIC set to :ada:`False`:
+
+.. code-block:: ada
+
+    package P is
+
+       type Q is limited private with
+         Default_Initial_Condition => False;
+
+       function F return Q;
+
+    private
+
+       type Q is range -1 .. 100;
+
+    end P;
+
+    package body P is
+
+       function F return Q is (42);
+
+    end P;
+
+    with Ada.Text_IO;  use Ada.Text_IO;
+
+    with P; use P;
+
+    procedure Main is
+       Obj1 : constant Q := F;
+       Obj2 : Q;  --  triggers Assertion_Error
+    begin
+       Put_Line (Obj1'Image);
+       Put_Line (Obj2'Image);
+       Put_Line ("Done");
+    end Main;
+
+In the above, :ada:`Assertion_Error` is raised by the elaboration of
+:ada:`Obj2` because the DIC check necessarily fails. There is no check on the
+declaration of :ada:`Obj1` because it is initialized, explicitly.
+
+
```

Comment:
Let's add the following here:

To recap, we can ensure initialization for objects of the type by detecting, during elaboration at run-time, any objects not explicitly initialized.

This approach is sufficient because when elaboration of an object declaration raises an exception, no use of that object is possible. That's guaranteed because the frame containing that declarative part is immediately abandoned and the exception is propagated up to the previous level. A local handler never can apply. But even if there is a matching handler in the previous level, there's really nothing much to be done. Re-entering the frame containing the declaration will raise the exception all over again, necessarily. Thus the code will have to be changed and recompiled, meeting the goal of the idiom.

We can illustrate this assurance using Storage_Error. Consider the following program, in which the main procedure calls an inner procedure P:

with Text_IO;  use Text_IO;

procedure Main is

   procedure P (Output : out Float) is
      N : array (Positive) of Float;   -- Storage_Error is likely
   begin
      Put_Line ("P's body assigns N's components and uses them");
      --  The following indexes and component values are arbitrary 
      --  and used purely for illustration...
      N := (others => 0.0);  
      --  other computations and assignments to N ...      
      Output := N (5);  
   exception
      when Storage_Error =>
         Output := N (1);
   end P;

   X : Float;
   
begin
   P (X);
   Put_Line (X'Image);   
   Put_Line ("Done");
exception
   when Storage_Error =>
      Put_Line ("Main completes abnormally");
end Main;

When Main calls P, the elaboration of the declarative part of P almost certainly fails because there is insufficient storage to allocate to the object P.N, hence Storage_Error is raised. (If your machine can handle the above, congratulations.) Even though procedure P has a handler specifically for Storage_Error, that  handler never applies because the declarative part is immediately abandoned. Instead, the exception is raised in the caller, where it can be caught. This behavior is essential to ensure that problematic objects are not referenced in the local handlers. In the above, the handler in P for Storage_Error references the object P.N to assign the P.Output parameter. If that assignment could happen -- again, it cannot -- what would it mean, functionally? No one knows.

Handling Storage_Error is a little tricky anyway. Does the OS give the program a chance to execute a handler? If so, is there sufficient storage remaining to execute the exception handler's statements? In any case you can see the problem that the declaration failure semantics preclude.

Therefore, although the DIC solution is not enforced at compile-time, it is nevertheless sufficient to ensure no uninitialized object of the type can be used.